src/xo-app/new/vm/index.js

```diff
@@ -1,214 +1,766 @@
+    this.setState({ pool })
+    this._reset()
+  }
+
+  _getIsInPool = createSelector(
+    () => this.state.pool,
+    pool => object => object.$pool === pool.id
+  )
+
+  _getSrPredicate = createSelector(
+    () => this.state.pool,
+    pool => disk => disk.$pool === pool.id && disk.content_type === 'user'
+  )
+
+  _initTemplate = template => {
+    if (!template) {
+      return this._reset()
+    }
+
+    const state = store.getState()
+
+    const existingDisks = {}
+    forEach(template.$VBDs, vbdId => {
+      const vbd = getObject(state, vbdId)
+      if (vbd.is_cd_drive) {
+        return
+      }
+      const vdi = getObject(state, vbd.VDI)
+      existingDisks[this.uniqueId] = {
+        name_label: vdi.name_label,
+        name_description: vdi.name_description,
+        size: vdi.size,
+        $SR: vdi.$SR
+      }
+    })
+
+    const VIFs = []
+    forEach(template.VIFs, vifId => {
+      const vif = getObject(state, vifId)
+      VIFs.push({
+        mac: vif.MAC,
+        network: vif.$network,
+        id: this.uniqueId
+      })
+    })
+
+    this.setState({
+      // infos
+      name_label: template.name_label,
+      template,
+      name_description: template.name_description || this.state.name_description || '',
+      // performances
+      memory: template.memory.size,
+      CPUs: template.CPUs.number,
+      cpuWeight: 1,
+      // installation
+      installMethod: template.install_methods && template.install_methods[0] || this.state.installMethod,
+      // interfaces
+      VIFs,
+      // disks
+      existingDisks,
+      VDIs: map(template.template_info.disks, disk => ({ ...disk, device: this.uniqueId }))
+    }, () => forEach(this.state, (element, key) => {
+      !isArray(element) && this._setRef(key, element)
+    }))
+  }
+
+  _addVdi = () => this.setState({ VDIs: concat(this.state.VDIs, {
+    device: String(this.uniqueId),
+    type: 'system'
+  }) })
+  _removeVdi = index => {
+    const VDIs = cloneDeep(this.state.VDIs)
+    pullAt(VDIs, index)
+    this.setState({ VDIs })
+  }
+  _addInterface = () => this.setState({ VIFs: concat(this.state.VIFs, {
+    id: this.uniqueId
+  }) })
+  _removeInterface = index => {
+    const VIFs = cloneDeep(this.state.VIFs)
+    pullAt(VIFs, index)
+    this.setState({ VIFs })
+  }
+
+  _getOnChange = (prop) => {
+    const debouncer = debounce(param => this.setState({ [prop]: param }), 100)
+    return param => {
+      const _param = param && param.target ? param.target.value : param
+      debouncer(_param)
+    }
+  }
+  _getOnChangeCheckbox = (prop) => {
+    const debouncer = debounce(event =>
+      this.setState({ [prop]: event.target.checked }), 100
+    )
+    return event => {
+      const _param = event.target.checked
+      debouncer(_param)
+    }
+  }
+  _getOnChangeObject = (stateElement, key, stateProperty, targetProperty) => {
+    const debouncer = debounce(param => {
+      const stateValue = cloneDeep(this.state[stateElement])
```

Comment:
At least do not clone deeply unless really necessary…

And you can do: `const clone = { ...obj }`.